### ISSUE TITLE
Support libbi dual tariff

### ISSUE THREAD

**Issue by plord12** [CJNE/pymyenergi#20]
It would be great if we could support libbi dual tariff ... for example, this would allow "charge from grid now" and "sync libbi tariff from dynamic tariff (such as Octopus Intelligent)"

API is 

```
GET https://myaccount.myenergi.com/api/EnergySetup/GetDualTariffEnergyPrices?hubId=<hubid>
```

returns 

```
{
    "content": [
        {
            "days": [
                0,
                1,
                2,
                3,
                4,
                5,
                6
            ],
            "energySetupId": "47a272a0-584d-ee11-abf4-0aa731bd900a",
            "tariffs": [
                {
                    "fromMinutes": 0,
                    "id": "3fc421ff-d283-ee11-abf4-0aa731bd900a",
                    "price": 7.5,
                    "toMinutes": 300
                },
                {
                    "fromMinutes": 1410,
                    "id": "40c421ff-d283-ee11-abf4-0aa731bd900a",
                    "price": 7.5,
                    "toMinutes": 1440
                },
                {
                    "fromMinutes": 360,
                    "id": "41c421ff-d283-ee11-abf4-0aa731bd900a",
                    "price": 30.6,
                    "toMinutes": 1410
                },
                {
                    "fromMinutes": 330,
                    "id": "42c421ff-d283-ee11-abf4-0aa731bd900a",
                    "price": 30.6,
                    "toMinutes": 360
                },
                {
                    "fromMinutes": 300,
                    "id": "43c421ff-d283-ee11-abf4-0aa731bd900a",
                    "price": 30.6,
                    "toMinutes": 330
                }
            ]
        }
    ],
    "field": "",
    "message": "",
    "status": true
}
```

then to update :

```
POST https://myaccount.myenergi.com/api/EnergySetup/SaveDualTariffEnergyPrices
```

with

```
[
    {
        "days": [
            0,
            1,
            2,
            3,
            4,
            5,
            6
        ],
        "energySetupId": "47a272a0-584d-ee11-abf4-0aa731bd900a",
        "tariffs": [
            {
                "fromMinutes": 0,
                "price": 7.5,
                "toMinutes": 300
            },
            {
                "fromMinutes": 1410,
                "price": 7.5,
                "toMinutes": 1440
            },
            {
                "fromMinutes": 360,
                "price": 30.6,
                "toMinutes": 1410
            },
            {
                "fromMinutes": 330,
                "price": 30.6,
                "toMinutes": 360
            },
            {
                "fromMinutes": 300,
                "price": 30.6,
                "toMinutes": 330
            }
        ]
    }
]
```

**Comment by plord12** [comment]
For a quick hack, I added this to my linux server -

```
$ cat set-libbi-rates.py
#!libbi-venv/bin/python3

import requests
import json
import sys
from pycognito import Cognito

u = Cognito('eu-west-2_E57cCJB20','2fup0dhufn5vurmprjkj599041', username='XXX')
u.authenticate(password='XXX')
headers = {"Authorization": f"Bearer {u.access_token}"}

r = requests.get("https://myaccount.myenergi.com/api/AccountAccess/Hubs", headers=headers)
hubid = json.loads(r.text).get('content').get('hubs')[0].get('hubId')
serial = json.loads(r.text).get('content').get('hubs')[0].get('serialNo')

r = requests.get("https://myaccount.myenergi.com/api/EnergySetup/GetDualTariffEnergyPrices?hubId="+hubid, headers=headers)
id = json.loads(r.text).get('content')[0].get('energySetupId')

# parse args :
#
# 00:00-05:30,7.5 05:30-23:30,30.6 23:30-00:00,7.5

jsonString='[ { "days": [0,1,2,3,4,5,6], "energySetupId": "'+id+'", "tariffs": ['

for i in sys.argv[1:]:
    time,price=i.split(",")
    start,end=time.split("-")
    fromMinutes=int(start.split(":")[0])*60+int(start.split(":")[1])
    toMinutes=int(end.split(":")[0])*60+int(end.split(":")[1])
    if toMinutes == 0:
        toMinutes = 1440
    jsonString=jsonString+'{ "fromMinutes": '+str(fromMinutes)+', "price": '+price+', "toMinutes": '+str(toMinutes)+' },'

data=json.loads(jsonString[:-1]+'] } ]')

r = requests.post("https://myaccount.myenergi.com/api/EnergySetup/SaveDualTariffEnergyPrices", headers=headers, json=data)
print(json.loads(r.text).get('message'))

$ cat set-libbi-rates.sh 
#!/bin/sh
#

cd $(dirname $0)

echo $* > /tmp/libbi.txt
./set-libbi-rates.py $* 2>&1 | tee /tmp/libbi.log
```

And then in home assistant -

```
shell_command:
  libbi_octopus_rates: >
    ssh plord@arm3.lan -i /config/.ssh/id_rsa -o UserKnownHostsFile=/config/.ssh/known_hosts 
    "/home/plord/bin/set-libbi-rates.sh "{%- set rates=state_attr('event.octopus_energy_electricity_XXXX_current_day_rates','rates') -%} 
    {%- for rate in rates -%} 
    {%- set adjusted_rate=rate.value_inc_vat*100 -%}
    {{- '{:02}:{:02}-{:02}:{:02},{:.02f} '.format(rate.start.hour, rate.start.minute,rate.end.hour, rate.end.minute, adjusted_rate) -}} 
    {%- endfor -%}
```

To sync libbi rates from my octopus rates.  Whats nice here is that when octopus creates boosts for intellegent go, the libi also charges up at the same time (cheap rates).

(lots of issues with this, of course, such as no error checking, assumes libbi has one tariff manually configured etc etc).

**Comment by plord12** [comment]
I'm wondering if this is useful now.  Once home assistant supports charge from grid then charging/export can be managed by home assistant.

**Comment by HLFCode** [comment]
> Once home assistant supports charge from grid then charging/export can be managed by home assistant.

Not sure what you mean by this, will you explain please.

AFAIUI HA uses pymyenergi, so setting anything in the libbi has to be provided by this repository doesn't it?

I really want to do what you're doing (charge the libbi when Go uses "day" times) but there's no myenergi api to say "charge now" is there?
If there isn't, the only way is to set a temporary tariff as you describe above.

**Comment by trizmark** [comment]
@plord12 I want to take a stab at implementing this. I know this has been opened for a *long* time, but with the Octopus free energy sessions, changing the schedule all the time is rather tedious.
Have you figured out how the id is generated for each time band? It looks like they take the second half of the energySetupId, but with a different prefix. Looks like they're consecutive IDs, but incrementing on the most significant part of the ID. (Weird way to do it)

**Comment by plord12** [comment]
> [@plord12](https://github.com/plord12) I want to take a stab at implementing this. I know this has been opened for a _long_ time, but with the Octopus free energy sessions, changing the schedule all the time is rather tedious. Have you figured out how the id is generated for each time band? It looks like they take the second half of the energySetupId, but with a different prefix. Looks like they're consecutive IDs, but incrementing on the most significant part of the ID. (Weird way to do it)

FYI, I no-longer do this.  

My home assistant takes signals from my energy provider, octopus, for the automations.  Eg when electricity cost is low, charge the libbi.  I've made the myenergi libbi schedule flat (to effectively disable controlling charging from myenergi).

**Comment by trizmark** [comment]
Fair enough. I had a similar setup to that (flat cheap tariff), but then once I had a failure to stop the charging and now I am back to the scheduled charge.
I'll have a go at implementing this anyway.

**Comment by plord12** [comment]
> > [@plord12](https://github.com/plord12) I want to take a stab at implementing this. I know this has been opened for a _long_ time, but with the Octopus free energy sessions, changing the schedule all the time is rather tedious. Have you figured out how the id is generated for each time band? It looks like they take the second half of the energySetupId, but with a different prefix. Looks like they're consecutive IDs, but incrementing on the most significant part of the ID. (Weird way to do it)
> 
> FYI, I no-longer do this.
> 
> My home assistant takes signals from my energy provider, octopus, for the automations. Eg when electricity cost is low, charge the libbi. I've made the myenergi libbi schedule flat (to effectively disable controlling charging from myenergi).

Having said that, I see myenergi have recently disabled our own control of charging from the grid and so I think we do need the ability to read and set the libbi tariff from homeasistant.

**Comment by trizmark** [comment]
You mean you can't control the 'charge from grid' from HA? It works for me....

**Comment by plord12** [comment]
> You mean you can't control the 'charge from grid' from HA? It works for me....

No-longer working for me ... my battery failed to charge overnight in the last 2 days.

Looks like myenergi only enable charge-from-grid tickbox now for certain tariffs ... eg choose a fixed rate tariff and you don't get the tick box in the official app.

<img width="781" height="116" alt="Image" src="https://github.com/user-attachments/assets/f4f0cdfb-a642-4b9e-bd73-706611ddbf1f" />

If you choose the right tariff you do get a tick box, but it doesn't always seem to work.

<img width="780" height="147" alt="Image" src="https://github.com/user-attachments/assets/6f3be66a-cf30-4592-ad94-25089b2d3db1" />

**Comment by MalcolmSpencer** [comment]
I have set my rates at low apart from 1 half hour period at 6-6:30pm when i am unlikely to charge from grid as prices are always high. This allows me to control the Libbi charging by toggling the 'charge from grid' in HA. This has been working fine for a while now.

**Comment by G6EJD** [comment]
Not in any form a derogatory way; I want to say at last, someone who is understanding control theory, because too high a refresh rate, certainly higher than the system under control, will cause what is in effect an oscillation. I'm sure you will now see continued stability in your system. 

I have seen discussion on the MyEnergi forum where some have their refresh set at 1-second intervals, but I had long since given up on explaining what the effect is.

**Comment by trizmark** [comment]
@G6EJD dude what? Nobody is talking about refresh rate here....  The discussion is about tariff handling. Pls read the comments before you chime in.

**Comment by G6EJD** [comment]
Well you should be
